service/udp.go

```diff
@@ -321,92 +314,231 @@ func (c *natconn) ReadFrom(buf []byte) (int, net.Addr, error) {
 // Packet NAT table
 type natmap struct {
 	sync.RWMutex
-	keyConn map[string]*natconn
-	logger  *slog.Logger
-	timeout time.Duration
-	metrics UDPMetrics
+	associations map[string]PacketAssociation
 }
 
-func newNATmap(timeout time.Duration, sm UDPMetrics, l *slog.Logger) *natmap {
-	m := &natmap{logger: l, metrics: sm}
-	m.keyConn = make(map[string]*natconn)
-	m.timeout = timeout
-	return m
+func newNATmap() *natmap {
+	return &natmap{associations: make(map[string]PacketAssociation)}
 }
 
-func (m *natmap) Get(key string) *natconn {
+// Get returns a UDP NAT entry from the natmap.
+func (m *natmap) Get(clientAddr string) PacketAssociation {
 	m.RLock()
 	defer m.RUnlock()
-	return m.keyConn[key]
+	return m.associations[clientAddr]
 }
 
-func (m *natmap) set(key string, pc net.PacketConn, cryptoKey *shadowsocks.EncryptionKey, connMetrics UDPConnMetrics) *natconn {
-	entry := &natconn{
-		PacketConn:     pc,
-		cryptoKey:      cryptoKey,
-		metrics:        connMetrics,
-		defaultTimeout: m.timeout,
+// Del deletes a UDP NAT entry from the natmap.
+func (m *natmap) Del(clientAddr string) {
+	m.Lock()
+	defer m.Unlock()
+
+	if _, ok := m.associations[clientAddr]; ok {
+		delete(m.associations, clientAddr)
 	}
+}
 
+// Add adds a new UDP NAT entry to the natmap.
+func (m *natmap) Add(clientAddr string, assoc PacketAssociation) {
 	m.Lock()
 	defer m.Unlock()
 
-	m.keyConn[key] = entry
-	return entry
+	m.associations[clientAddr] = assoc
 }
 
-func (m *natmap) del(key string) net.PacketConn {
+func (m *natmap) Close() error {
 	m.Lock()
 	defer m.Unlock()
 
-	entry, ok := m.keyConn[key]
-	if ok {
-		delete(m.keyConn, key)
-		return entry
+	var err error
+	for _, assoc := range m.associations {
+		if e := assoc.Close(); e != nil {
+			err = e
+		}
+	}
+	return err
+}
+
+func HandleAssociation(assoc PacketAssociation) { 
+	for {
+		lazySlice := readBufPool.LazySlice()
+		buf := lazySlice.Acquire()
+		n, err := assoc.Read(buf)
+		if errors.Is(err, net.ErrClosed) {
+			lazySlice.Release()
+			return
+		}
+		pkt := buf[:n]
+		select {
+		case <-assoc.Done():
+			lazySlice.Release()
+			return
+		default:
+			go assoc.HandlePacket(pkt, lazySlice)
+		}
+	}
+}
+
+// PacketAssociation represents a UDP association that handles individual packets.
+type PacketAssociation interface {
+	// HandlePacket processes a single incoming packet.
+	//
+	// pkt contains the raw packet data.
+	// lazySlice is the LazySlice that holds the pkt buffer, which should be
+	// released after the packet is processed.
+	HandlePacket(pkt []byte, lazySlice slicepool.LazySlice)
```

Comment:
This is confusing. The association and the packet handling should be decoupled. We should be able to use different implementations to handle an association.